### PR TITLE
SANSILLAutoProcess select stitching reference

### PR DESCRIPTION
**Pull request by darominski**
A new property, `IndexOfReference` has been added to `Stitch1DMany` algorithm. It allows user to specify which of the input workspaces will provide the scale for the stitched output. To communicate with `Stitch1DMany` which workspace has been selected, a new property has been added to `SANSILLAutoProcess`, called `StitchReferenceIndex`, that is passed to the stitching algorithm. 

**To test:**
1. Using a slightly modified example from Stitch1DMany documentation, run the following script in the workbench:

```
def gaussian(x, mu, sigma):
  """Creates a gaussian peak centered on mu and with width sigma."""
  return (1/ sigma * np.sqrt(2 * np.pi)) * np.exp( - (x-mu)**2  / (2*sigma**2))

# Create six histograms with a single peak in each one
x1 = np.arange(-1, 1, 0.02)
x3 = np.arange(0.3, 1.8, 0.02)
x5 = np.arange(1.4, 2.8, 0.02)
ws1 = CreateWorkspace(UnitX="1/q", DataX=x1, DataY=gaussian(x1[:-1], 0, 0.1)+1)
ws3 = CreateWorkspace(UnitX="1/q", DataX=x3, DataY=gaussian(x3[:-1], 1, 0.05)+1)
ws5 = CreateWorkspace(UnitX="1/q", DataX=x5, DataY=gaussian(x5[:-1], 2, 0.12)+1)

stitched_0 = Stitch1DMany(InputWorkspaces=[ws1,ws3,ws5], IndexOfReference=0)
stitched_1 = Stitch1DMany(InputWorkspaces=[ws1,ws3,ws5], IndexOfReference=1)
stitched_2 = Stitch1DMany(InputWorkspaces=[ws1,ws3,ws5], IndexOfReference=2)
```
2. Verify that in the output of stitching, the scale of the workspace used as reference has not changed, e.g. overplot w1 with stitched_0.

Part of #30343 

Release notes added.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gui-co** [comment]
The provided test works. I have also tested the new parameter in SANSILLAutoProcess which works as expected.

The only problem I see is that there is nothing to prevent a negative index from being provided.

Related to this, since the index is always positive, you could change its type to `size_t` (instead of `int`). This would allow you to remove all `static_cast` you do on vector sizes. Here this is not a problem because the vectors will never be that big, but this kind of conversion (`unsigned int` -> `int`) can be dangerous.

**Comment by darominski** [comment]
> The provided test works. I have also tested the new parameter in SANSILLAutoProcess which works as expected.
> 
> The only problem I see is that there is nothing to prevent a negative index from being provided.
> 
> Related to this, since the index is always positive, you could change its type to `size_t` (instead of `int`). This would allow you to remove all `static_cast` you do on vector sizes. Here this is not a problem because the vectors will never be that big, but this kind of conversion (`unsigned int` -> `int`) can be dangerous.

Thanks for testing and the suggestion. Indeed, the validation on the C++ algorithm side was missing, while it was there for the `SANSILLAutoProcess`, but that is definitely insufficient for other possible workflows. The changed the type of the `indexOfReference` to `size_t`, as suggested.

**Review by gui-co — DISMISSED** [review]

File: Framework/Algorithms/src/Stitch1DMany.cpp [hunk 1]
```diff
@@ -275,11 +287,19 @@ void Stitch1DMany::doStitch1D(std::vector<MatrixWorkspace_sptr> &toStitch,
 
   auto lhsWS = toStitch.front();
   outName += "_" + lhsWS->getName();
+  auto scaleRHSWorkspace = m_scaleRHSWorkspace;
 
   for (size_t i = 1; i < toStitch.size(); i++) {
-
     auto rhsWS = toStitch[i];
     outName += "_" + rhsWS->getName();
+    if (i >= m_indexOfReference) {
```
Comment (gvardany) [inline Framework/Algorithms/src/Stitch1DMany.cpp:295]: This branching can be made one level less.

File: Framework/Algorithms/test/Stitch1DManyTest.h [hunk 2]
```diff
@@ -388,6 +388,7 @@ class Stitch1DManyTest : public CxxTest::TestSuite {
     alg.setProperty("StartOverlaps", "0.8, 1.6");
     alg.setProperty("EndOverlaps", "1.1, 1.8");
     alg.setProperty("OutputWorkspace", "outws");
+    alg.setProperty("IndexOfReference", "0");
```
Comment (gvardany) [inline Framework/Algorithms/test/Stitch1DManyTest.h:391]: If 0 is the default, then you shouldn't need to specify here.
In fact, it makes a good test: if you don't specify it, does the test still pass?
This ensures backwards compatibility.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py [hunk 3]
```diff
@@ -405,7 +410,9 @@ def PyExec(self):
             try:
                 stitched = self.output + "_stitched"
                 Stitch1DMany(InputWorkspaces=outputSamples,
-                             OutputWorkspace=stitched)
+                             OutputWorkspace=stitched,
+                             ScaleRHSWorkspace=True,
+                             indexOfReference=self.stitch_reference_index)
```
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:415]: The key must be start uppercase here.

File: docs/source/algorithms/Stitch1DMany-v1.rst [hunk 4]
```diff
@@ -40,6 +40,9 @@ We can specify manual scale factors to use by setting
 of scale factors from the selected period. These scale factors are then applied
 to all other periods when stitching.
 
+The workspace that provides the scale for stitching output can be chosen by
+specifying the desired index through :literal:`IndexOfReference` property.
```
Comment (gvardany) [inline docs/source/algorithms/Stitch1DMany-v1.rst:44]: Maybe clarify: The reference workspace will not be scaled; the other will be scaled to match the reference.

File: docs/source/release/v6.1.0/framework.rst [hunk 5]
```diff
@@ -14,16 +14,14 @@ Concepts
 
 Algorithms
 ----------
-:ref:`LoadNexusLogs <algm-LoadNexusLogs>` has additional parameters to allow or block specific logs from being loaded.
```
Comment (gvardany) [inline docs/source/release/v6.1.0/framework.rst:17]: What happened to these 2 notes?

File: docs/source/release/v6.1.0/sans.rst [hunk 6]
```diff
@@ -18,5 +19,18 @@ Improvements
...
+  `StitchReferenceIndex` to denote the index of ws that should be a reference
+  for scaling during stitching
+
+Algorithms and instruments
+--------------------------
+
+Improvements
+############
+
+- :ref:`SANSILLAutoProcess <algm-SANSILLAutoProcess>` has new property:
```
Comment (gvardany) [inline docs/source/release/v6.1.0/sans.rst:32]: This note is duplicated above. Also add a note for Stitch1DMany itself in framework section.

**Review by gvardany — COMMENTED** [review]
Few minor issues spotted:

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py (hunk 3 above)
Comment (gvardany) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:415]: Besides, there is another Stitch at line 460 for the wedges.

File: Framework/Algorithms/test/Stitch1DManyTest.h (hunk 2 above)
Comment (darominski) [reply Framework/Algorithms/test/Stitch1DManyTest.h:391]: It absolutely passes, I'll remove that.

File: docs/source/release/v6.1.0/framework.rst (hunk 5 above)
Comment (darominski) [reply docs/source/release/v6.1.0/framework.rst:17]: I see that resolving conflicts and merging the release notes went pretty poorly. I'll correct it.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py (hunk 3 above)
Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:415]: Currently, they are going to be stitched by default, with the reference being the first one. Would you like to use the same `stitch_reference_index` for both sample data and wedges?

File: docs/source/release/v6.1.0/sans.rst (hunk 6 above)
Comment (darominski) [reply docs/source/release/v6.1.0/sans.rst:32]: There is a note for Stitch1DMany in the framework section already, also poorly merged, with a missing line, which naturally will be corrected.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py (hunk 3 above)
Comment (gvardany) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:415]: Yes.

Comment (gvardany) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:415]: Also, for both cases, make sure the stitch is attempted only if there are more than one workspaces to stitch. Especially now that the default reference index is 1.

File: Framework/Algorithms/src/Stitch1DMany.cpp [hunk 7]
```diff
@@ -76,6 +76,11 @@ void Stitch1DMany::init() {
       std::make_unique<VisibleWhenProperty>(useManualScaleFactorsTrue, manualScaleFactorsDefault, AND);
 
   setPropertySettings("ScaleFactorFromPeriod", std::move(scaleFactorFromPeriodVisible));
+
+  auto mustBePositive = std::make_shared<BoundedValidator<size_t>>();
+  mustBePositive->setLower(0);
+  declareProperty(std::make_unique<PropertyWithValue<size_t>>("IndexOfReference", 0, mustBePositive, Direction::Input),
```
Comment (gvardany) [inline Framework/Algorithms/src/Stitch1DMany.cpp:82]: The algorithms can be called from python, which doesn't understand unsigned.
So size_t should be avoided in the properties. Use int but keep the validator.
Also for primitives, you can use the simplified override of declareProperty, i.e. you don't have to make an explicit PropertyWithValue<T>.

**Review by gvardany — COMMENTED** [review]
Also please don't forget to expose the new input in Drill global settings.

File: scripts/Interface/ui/drill/model/configurations.py [hunk 8]
```diff
@@ -221,7 +221,8 @@ class RundexSettings(object):
                 "DeltaQ",
                 "IQxQyLogBinning",
                 "OutputPanels",
-                "WavelengthRange"
+                "WavelengthRange",
+                "StitchReferenceIndex"
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/configurations.py:225]: Thanks for the changes, one more small request, although unrelated.
Could you also add the ClearCorrected2DWorkspace here?

Comment (darominski) [reply scripts/Interface/ui/drill/model/configurations.py]: Yes, done.

**Comment by gvardany** [comment]
Thanks, @gui-co could you reapprove?

**Review by gui-co — APPROVED** [review]

**Pull request merged**